libs/platform/user/kernel_um.cpp

```diff
@@ -2,18 +2,147 @@
 // SPDX-License-Identifier: MIT
 
 #include <condition_variable>
+#include <map>
 #include <mutex>
+#include <tuple>
 
 #include "ebpf_platform.h"
 #include "kernel_um.h"
 
-typedef struct _mock_rundown_ref
-{
-    std::mutex lock;
-    std::condition_variable cv;
-    size_t count = 0;
-    bool rundown_in_progress = false;
-} mock_rundown_ref;
+/***
+ * @brief This following class implements a mock of the Windows Kernel's rundown reference implementation.
+ * 1) It uses a map to track the number of references to a given EX_RUNDOWN_REF structure.
+ * 2) The address of the EX_RUNDOWN_REF structure is used as the key to the map.
+ * 3) A single condition variable is used to wait for the ref count of any EX_RUNDOWN_REF structure to reach 0.
+ * 4) The class is a singleton and is created during static initialization and destroyed during static destruction.
+ */
+typedef class _rundown_ref_table
+{
+  private:
+    static std::unique_ptr<_rundown_ref_table> _instance;
+
+  public:
+    // The constructor and destructor should be private to ensure that the class is a singleton, but that is not
+    // possible because the singleton instance is stored in a unique_ptr which requires the constructor to be public.
```

Comment:
Not sure if this will help, but there seem to be a couple of approaches that fix/workarond the 'unauthorized use of public constructors' scenario for std::unique_ptrs: https://seanmiddleditch.github.io/enabling-make-unique-with-private-constructors/

There's also a pretty interesting discussion @ https://stackoverflow.com/questions/50995599/make-unique-doesnt-compile-for-creating-a-singleton-instance

ofc, you may already know of these workarounds and found them unsuitable for some reasons I'm quite possibly overlooking :-P

Reply:
All of the workarounds I saw rely on undefined behavior in the STL libraries. They make assumptions about implementation details of std::unique_ptr.